File: backend/app/schemas/generation.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.aliases import to_camel
from app.schemas.mockup import MockupGenerationPlan
from app.schemas.product import ProductInfo
from app.schemas.prompt_config import PromptConfig, StyleMemorySelection


class CreateGenerationTaskRequest(BaseModel):
    product_id: str | None = None
    images: list[str]
    platform: str
    country: str
    language: str
    quality_level: str
    image_ratio: str | None = None
    image_model: str | None = None
    design_style: str | None = None
    product_info: ProductInfo
    modules: list[str]
    prompt_config: PromptConfig = Field(default_factory=PromptConfig)
    style_memory: StyleMemorySelection = Field(default_factory=StyleMemorySelection)
    mockup_plan: MockupGenerationPlan | None = None

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class GenerationTaskResponse(BaseModel):
    id: str
    product_name: str
    status: str
    progress: int
    current_step: str
    created_at: str
    error_message: str | None = None
    conversation_session_id: str | None = None

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class GeneratedModule(BaseModel):
    id: str
    type: str
    title: str
    subtitle: str | None = None
    description: str | None = None
    image_url: str | None = None
    visual_prompt: str | None = None
    layout: str
    sort_order: int

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class GenerationResultResponse(BaseModel):
    id: str
    task_id: str
    product_id: str
    modules: list[GeneratedModule]
    preview_url: str | None = None
    export_urls: dict[str, str] = Field(default_factory=dict)
    quality_score: float
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class CreateResultVersionRequest(BaseModel):
    label: str | None = None

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class ResultVersionResponse(BaseModel):
    id: str
    result_id: str
    version: int
    label: str
    created_at: str
    modules: list[GeneratedModule]
    preview_url: str | None = None
    export_urls: dict[str, str] = Field(default_factory=dict)
    quality_score: float
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class ResultVersionListResponse(BaseModel):
    items: list[ResultVersionResponse]

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

File: frontend/src/types/generation.ts
import type { ProductInfo } from './product'
import type { MockupGenerationPlan } from './mockup'

export interface GenerationSettings {
  platform: string
  country: string
  language: string
  qualityLevel: string
  imageRatio: string
  imageModel: string
  designStyle: string
}

export interface PromptConfig {
  visualStyle: string[]
  color: string[]
  composition: string[]
  copyTone: string[]
  content: string[]
  platform: string[]
  negative: string[]
}

export type StyleMemoryMode = 'none' | 'user' | 'company' | 'brand'

export interface StyleMemorySelection {
  mode: StyleMemoryMode
  memoryId?: string
  updateAfterGeneration: boolean
}

export interface PromptOption {
  key: string
  label: string
  prompt: string
  negative: boolean
}

export interface PromptOptionGroup {
  key: keyof PromptConfig // 等价于 key: 'visualStyle' | 'color' | 'composition' | 'copyTone' | 'content' | 'platform' | 'negative'
  label: string
  selection: 'single' | 'multiple'
  options: PromptOption[]
}

export interface CreateGenerationTaskPayload {
  productId?: string
  images: string[]
  platform: string
  country: string
  language: string
  qualityLevel: string
  imageRatio: string
  imageModel: string
  designStyle: string
  productInfo: ProductInfo
  modules: string[]
  promptConfig: PromptConfig
  styleMemory: StyleMemorySelection
  mockupPlan?: MockupGenerationPlan
}

export interface GenerationTask {
  id: string
  // MVP 后端的 agent run 不一定返回产品名，前端展示时允许为空。
  productName?: string
  status: 'queued' | 'running' | 'completed' | 'failed' | 'cancelled' | string
  progress: number
  currentStep: string
  createdAt: string
  errorMessage?: string
  conversationSessionId?: string
  // 新后端 result 嵌在 agent run 中，这里记录 resultId 方便后续扩展。
  resultId?: string
}

export interface GenerationProgressStep {
  key: string
  label: string
  detail: string
  percentage: number
}
